README.md
# PR-DC Arduino Library for Analog Devices AD7797

The AD7797 is a complete analog front end for high precision bridge sensor applications such as weigh scales. It contains a 24-bit sigma-delta (Σ-Δ) analog-to-digital converter (ADC). The on-chip instrumentation amplifier has a fixed gain of 128, allowing small amplitude signals such as those from bridge sensors to be directly interfaced to the ADC.

For more information please visit: https://www.analog.com/en/products/ad7797.html

## Requirements
[Arduino](https://www.arduino.cc/)<br>

This library is tested with
**Arduino IDE 1.8.16**

## Installation
Clone the repository to the Arduino libraries folder.

For example, on the Windows OS path is:
```
C:\Users\{User}\Documents\Arduino\libraries
```

For more information please visit: https://www.arduino.cc/en/guide/libraries#toc5

## Usage

For usage please check provided examples. 

## License
Copyright (C) 2022 PR-DC <info@pr-dc.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as 
published by the Free Software Foundation, either version 3 of the 
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.

CHANGES.txt
--------------------
PRDC_AD7797
Analog Devices AD7797 Arduino Library
--------------------

Release 1.0.1, 25.01.2022.

- Arduino library specification

--------------------

Release 1.0.0, 06.01.2022.

- Initial release

--------------------


src/PRDC_AD7797.cpp
/**
 * PRDC_AD7797.cpp - Analog Devices AD7797 ADC Library
 * Author: Veljko Petrovic <vpetrovic@pr-dc.com>
 * PR-DC, Republic of Serbia
 * info@pr-dc.com
 *
 * --------------------
 * Copyright (C) 2021 PR-DC <info@pr-dc.com>
 *
 * This file is part of PRDC_AD7797.
 *
 * PRDC_AD7797 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * PRDC_AD7797 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with PRDC_AD7797.  If not, see <https://www.gnu.org/licenses/>.
 *
 */


 #include "PRDC_AD7797.h"

 // PRDC_AD7797()
 // Object constructor
 // --------------------
 PRDC_AD7797::PRDC_AD7797() :
   _spiSettings(AD7797_DEFAULT_SPI_FREQUENCY, MSBFIRST, SPI_MODE3),
   _spi(&AD7797_DEFAULT_SPI), _CS(AD7797_DEFAULT_CS), _MISO(AD7797_DEFAULT_MISO) {

 }

 // setSPI() function
 // Set SPI object
 // --------------------
 void PRDC_AD7797::setSPI(SPIClass& spi) {
   #ifdef DEBUG_AD7797
     Serial.println(F("setSPI()"));
   #endif

   _spi = &spi;
 }

 // setSPIFrequency() function
 // Set SPI frequency
 // --------------------
 void PRDC_AD7797::setSPIFrequency(uint32_t frequency) {
   #ifdef DEBUG_AD7797
     Serial.println(F("setSPIFrequency()"));
   #endif

   _spiSettings = SPISettings(frequency, MSBFIRST, SPI_MODE3);
 }

 // beginTransaction() function
 // Begin SPI transaction
 // --------------------
 inline void PRDC_AD7797::beginTransaction() {
   _spi->beginTransaction(_spiSettings);
   digitalWrite(_CS, LOW);
 }

 // endTransaction() function
 // End SPI transaction
 // --------------------
 inline void PRDC_AD7797::endTransaction() {
   digitalWrite(_CS, HIGH);
   _spi->endTransaction();
 }

 // getRegister() function
 // Reads the value of a register
 // --------------------
 uint32_t PRDC_AD7797::getRegister(uint8_t registerAddress, uint8_t bytesNumber) {
   #ifdef DEBUG_AD7797
     Serial.println(F("getRegister()"));
     Serial.print(F("Register "));
     Serial.println(registerAddress);
   #endif

   uint32_t buffer = 0x0;
   _spi->transfer(AD7797_COMM_READ | AD7797_COMM_ADDR(registerAddress));
   for(uint8_t i = 0; i < bytesNumber; i++)  {
       buffer = (buffer << 8) + _spi->transfer(0);
   }

   #ifdef DEBUG_AD7797
     Serial.println(buffer);
   #endif
   return buffer;
 }

 // getSingleRegister() function
 // Reads the value of a single register
 // --------------------
 uint32_t PRDC_AD7797::getSingleRegister(uint8_t registerAddress, uint8_t bytesNumber) {
   #ifdef DEBUG_AD7797
     Serial.println(F("getSingleRegister()"));
   #endif

   this->beginTransaction();
   uint32_t regVal = this->getRegister(registerAddress, bytesNumber);
   this->endTransaction();
   return regVal;
 }

 // setRegister() function
 // Set the value of a register
 // --------------------
 void PRDC_AD7797::setRegister(uint8_t registerAddress, uint32_t value, uint8_t bytesNumber) {
   #ifdef DEBUG_AD7797
     Serial.println(F("setRegister()"));
     Serial.print(F("Register "));
     Serial.print(registerAddress);
     Serial.print(F(" Value "));
     Serial.println(value);
   #endif

   uint8_t* dataPointer = (uint8_t*)&value;
   _spi->transfer(AD7797_COMM_WRITE | AD7797_COMM_ADDR(registerAddress));
   for(uint8_t i = 0; i < bytesNumber; i++)  {
     _spi->transfer(*(dataPointer + (bytesNumber-1) - i));
   }
 }

 // setSingleRegister() function
 // Set the value of a single register
 // --------------------
 void PRDC_AD7797::setSingleRegister(uint8_t registerAddress, uint32_t value, uint8_t bytesNumber) {
   #ifdef DEBUG_AD7797
     Serial.println(F("setSingleRegister()"));
   #endif

   this->beginTransaction();
   this->setRegister(registerAddress, value, bytesNumber);
   this->endTransaction();
 }

 // begin() function
 // Begin communication
 // --------------------
 bool PRDC_AD7797::begin() {
   #ifdef DEBUG_AD7797
     Serial.println(F("begin()"));
     Serial.println(F("PRDC_AD7797 constructor instantiated successfully"));
     Serial.print(F("CS "));
     Serial.println(_CS);
     Serial.print(F("MISO "));
     Serial.println(_MISO);
   #endif

   _spi->begin();
   this->pinInit();
   this->reset();
   if(this->checkID()) {
     return true;
   }
   return false;
 }

 bool PRDC_AD7797::begin(uint8_t CS, uint8_t MISO) {
   _CS = CS;
   _MISO = MISO;
   return this->begin();
 }

 // end() function
 // End communication
 // --------------------
 void PRDC_AD7797::end() {
   #ifdef DEBUG_AD7797
     Serial.println(F("end()"));
   #endif

   _spi->end();
 }

 // pinInit() function
 // Pin initialization
 // --------------------
 void PRDC_AD7797::pinInit() {
   #ifdef DEBUG_AD7797
     Serial.println(F("pinInit()"));
   #endif

   pinMode(_CS, OUTPUT);
   digitalWrite(_CS, HIGH);
 }

 // reset() function
 // Reset ADC
 // --------------------
 void PRDC_AD7797::reset() {
   #ifdef DEBUG_AD7797
     Serial.println(F("reset()"));
   #endif

   this->beginTransaction();
   _spi->transfer(0xFF);
   _spi->transfer(0xFF);
   _spi->transfer(0xFF);
   _spi->transfer(0xFF);
   this->endTransaction();
   delay(1);
 }

 // setClockMode() function
 // Set clock mode (internal or external)
 // --------------------
 void PRDC_AD7797::setClockMode(uint8_t clock_mode) {
   #ifdef DEBUG_AD7797
     Serial.println(F("setClockMode()"));
   #endif

   _clock_mode = clock_mode;
 }

 // setRate() function
 // Set output data rate
 // --------------------
 void PRDC_AD7797::setRate(uint32_t rate) {
   #ifdef DEBUG_AD7797
     Serial.println(F("setRate()"));
   #endif

   _rate = rate;
 }

 // checkID() function
 // Check ADC ID register
 // --------------------
 bool PRDC_AD7797::checkID() {
   #ifdef DEBUG_AD7797
     Serial.println(F("checkID()"));
   #endif

   uint8_t regVal = getSingleRegister(AD7797_REG_ID, 1);
   #ifdef DEBUG_AD7797
     Serial.print(F("ID = "));
     Serial.println(regVal & AD7797_ID_MASK);
   #endif
   return (regVal & AD7797_ID_MASK) == AD7797_ID;
 }

 // waitReady() function
 // Waits for MISO pin to go low
 // --------------------
 void PRDC_AD7797::waitReady() {
   #ifdef DEBUG_AD7797
     Serial.println(F("waitReady()"));
   #endif
   while(digitalRead(_MISO)) {}
 }

 // setPower() function
 // Set device to idle or power-down
 // mode =
 //  0 - power-down
 //  1 - idle
 // --------------------
 void PRDC_AD7797::setPower(uint8_t mode) {
   #ifdef DEBUG_AD7797
     Serial.println(F("setPower()"));
   #endif

    uint32_t oldPwrMode = this->getSingleRegister(AD7797_REG_MODE, 2);
    oldPwrMode &= ~(AD7797_MODE_SEL(0x7));
    uint32_t newPwrMode = oldPwrMode |
                          AD7797_MODE_SEL((mode * (AD7797_MODE_IDLE)) |
                          (!mode * (AD7797_MODE_PWRDN)));
    this->setSingleRegister(AD7797_REG_MODE, newPwrMode, 3);
 }

 // channelSelect() function
 // Selects the channel to be enabled
 // --------------------
 void PRDC_AD7797::channelSelect(uint8_t channel) {
   #ifdef DEBUG_AD7797
     Serial.println(F("channelSelect()"));
   #endif
   uint32_t regValue = this->getSingleRegister(AD7797_REG_CONF, 2);
   regValue &= ~AD7797_CONF_CHAN(0xFF);
   regValue |= AD7797_CONF_CHAN(channel);
   this->setSingleRegister(AD7797_REG_CONF, regValue, 2);
 }

 // calibrate() function
 // Performs the given calibration to the specified channel
 // --------------------
 void PRDC_AD7797::calibrate(uint8_t mode, uint8_t channel) {
   #ifdef DEBUG_AD7797
     Serial.println(F("calibrate()"));
   #endif

   this->channelSelect(channel);
   uint32_t regValue = this->getSingleRegister(AD7797_REG_MODE, 2);
   regValue &= ~AD7797_MODE_SEL(0x7);
   regValue |= AD7797_MODE_SEL(mode);
   this->beginTransaction();
   this->setRegister(AD7797_REG_MODE, regValue, 2);
   this->waitReady();
   this->endTransaction();
 }

 // setPolarity() function
 // Selects the polarity of the coding
 // --------------------
 void PRDC_AD7797::setPolarity(uint8_t polarity) {
   #ifdef DEBUG_AD7797
     Serial.println(F("setPolarity()"));
   #endif

   uint32_t regValue = this->getSingleRegister(AD7797_REG_CONF, 2);
   regValue &= ~AD7797_CONF_UNIPOLAR;
   regValue |= polarity * AD7797_CONF_UNIPOLAR;
   this->setSingleRegister(AD7797_REG_CONF, regValue, 2);
   _polarity = polarity;
 }

 // singleConversion() function
 // Returns the result of a single conversion
 // --------------------
 uint32_t PRDC_AD7797::singleConversion() {
   #ifdef DEBUG_AD7797
     Serial.println(F("singleConversion()"));
   #endif

   uint32_t command = AD7797_MODE_SEL(AD7797_MODE_SINGLE) |
                      AD7797_MODE_CLKSRC(_clock_mode) |
                      AD7797_MODE_RATE(_rate);
   _mode = AD7797_MODE_SINGLE;
   this->beginTransaction();
   this->setRegister(AD7797_REG_MODE, command, 2);
   this->waitReady();
   uint32_t raw = this->getRegister(AD7797_REG_DATA, 3);
   this->endTransaction();

   return raw;
 }

 // continuousConversion() function
 // Returns the result of continuous conversion
 // --------------------
 uint32_t PRDC_AD7797::continuousConversion() {
   #ifdef DEBUG_AD7797
     Serial.println(F("continuousConversion()"));
   #endif
   if(_mode != AD7797_MODE_CONT){
     uint32_t command = AD7797_MODE_SEL(AD7797_MODE_CONT) |
                        AD7797_MODE_CLKSRC(_clock_mode) |
                        AD7797_MODE_RATE(_rate);
     _mode = AD7797_MODE_CONT;
     this->beginTransaction();
     this->setRegister(AD7797_REG_MODE, command, 2);
     this->waitReady();
     uint32_t raw = this->getRegister(AD7797_REG_DATA, 3);
     this->endTransaction();
     return raw;
   }
   else{
     this->beginTransaction();
     this->waitReady();
     uint32_t raw = this->getRegister(AD7797_REG_DATA, 3);
     this->endTransaction();
     return raw;
   }
 }

 // continuousReadAverage() function
 // Returns the average of several conversion results
 // --------------------
 uint32_t PRDC_AD7797::continuousReadAverage(uint32_t sampleNumber) {
   #ifdef DEBUG_AD7797
     Serial.println(F("continuousReadAverage()"));
   #endif

   uint32_t samplesAverage = 0;
   uint32_t command = AD7797_MODE_SEL(AD7797_MODE_CONT) |
                      AD7797_MODE_CLKSRC(_clock_mode) |
                      AD7797_MODE_RATE(_rate);

   this->beginTransaction();
   this->setRegister(AD7797_REG_MODE, command, 2);
   for(uint32_t i = 0; i < sampleNumber; i++) {
     this->waitReady();
     samplesAverage += this->getRegister(AD7797_REG_DATA, 3);
   }
   this->endTransaction();

   return samplesAverage/sampleNumber;
 }

 // // temperatureRead() function
 // // Read data from temperature sensor and converts it to Celsius degrees
 // // --------------------
 // float PRDC_AD7797::temperatureRead() {
 //   #ifdef DEBUG_AD7797
 //     Serial.println(F("temperatureRead()"));
 //   #endif
 //
 //   this->setPolarity(0); // Bipolar operation
 //   this->channelSelect(AD7797_CH_TEMP);
 //   uint32_t dataReg = this->singleConversion();
 //   dataReg -= 0x800000;
 //   return (float)dataReg/2815.0-273; // Celsius Temperature
 // }

 // rawToVolts() function
 // Converts 24-bit raw data to volts
 // --------------------
 // float PRDC_AD7797::rawToVolts(uint32_t raw, float vRef) {
 //   #ifdef DEBUG_AD7797
 //     Serial.println(F("rawToVolts()"));
 //   #endif
 //
 //   float voltage;
 //   if(_polarity == 0) {
 //     // Bipolar mode
 //     voltage = (((float)raw/(1ul << 23))-1)*vRef/_gain;
 //   } else {
 //     // Unipolar mode
 //     voltage = ((float)raw*vRef)/(1ul << 24)/_gain;
 //   }
 //
 //   return voltage;
 // }

 // printAllRegisters() function
 // Print value of each register
 // --------------------
 void PRDC_AD7797::printAllRegisters(void)  {
   Serial.println(F("Read all registers"));
   this->beginTransaction();
   Serial.println(F("Register 0"));
   Serial.println(this->getRegister(0, 1));
   Serial.println(F("Register 1"));
   Serial.println(this->getRegister(1, 3));
   Serial.println(F("Register 2"));
   Serial.println(this->getRegister(2, 3));
   Serial.println(F("Register 3"));
   Serial.println(this->getRegister(3, 3));
   Serial.println(F("Register 4"));
   Serial.println(this->getRegister(4, 1));
   Serial.println(F("Register 5"));
   Serial.println(this->getRegister(5, 1));
   Serial.println(F("Register 6"));
   Serial.println(this->getRegister(6, 3));
   Serial.println(F("Register 7"));
   Serial.println(this->getRegister(7, 3));
   this->endTransaction();
 }


src/PRDC_AD7797.h
/**
 * PRDC_AD7797.h - Analog Devices AD7797 ADC Library
 * Author: Veljko Petrovic <vpetrovic@pr-dc.com>
 * PR-DC, Republic of Serbia
 * info@pr-dc.com
 *
 * --------------------
 * Copyright (C) 2021 PR-DC <info@pr-dc.com>
 *
 * This file is part of PRDC_AD7797.
 *
 * PRDC_AD7797 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * PRDC_AD7797 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with PRDC_AD7797.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

#ifndef __PRDC_AD7797_H__
#define __PRDC_AD7797_H__

#include "Arduino.h"
#include <SPI.h>

// AD7797 Library debug
// #define DEBUG_AD7797

// SPI communication settings
#define AD7797_DEFAULT_SPI_FREQUENCY 1000000
#define AD7797_DEFAULT_SPI SPI
#define AD7797_DEFAULT_CS PIN_SPI_SS
#define AD7797_DEFAULT_MISO PIN_SPI_MISO

// AD7797 Register Map
#define AD7797_REG_COMM		0 // Communications Register(WO, 8-bit)
#define AD7797_REG_STAT	    0 // Status Register	    (RO, 8-bit)
#define AD7797_REG_MODE	    1 // Mode Register	     	(RW, 16-bit
#define AD7797_REG_CONF	    2 // Configuration Register (RW, 16-bit)
#define AD7797_REG_DATA	    3 // Data Register	     	(RO, 16-/24-bit)
#define AD7797_REG_ID	    4 // ID Register	     	(RO, 8-bit)
#define AD7797_REG_OFFSET   6 // Offset Register	    (RW, 24-bit)
#define AD7797_REG_FULLSALE	7 // Full-Scale Register	(RW, 24-bit)

// Communications Register Bit Designations (AD7797_REG_COMM)
#define AD7797_COMM_WEN		(1 << 7) 			// Write Enable
#define AD7797_COMM_WRITE	(0 << 6) 			// Write Operation
#define AD7797_COMM_READ    (1 << 6) 			// Read Operation
#define AD7797_COMM_ADDR(x)	(((x) & 0x7) << 3)	// Register Address
#define AD7797_COMM_CREAD	(1 << 2) 			// Continuous Read of Data Register

// Status Register Bit Designations (AD7797_REG_STAT)
#define AD7797_STAT_RDY		(1 << 7) // Ready
#define AD7797_STAT_ERR		(1 << 6) // Error (Overrange, Underrange)
#define AD7797_STAT_CH3		(1 << 2) // Channel 3
#define AD7797_STAT_CH2		(1 << 1) // Channel 2
#define AD7797_STAT_CH1		(1 << 0) // Channel 1

// Mode Register Bit Designations (AD7797_REG_MODE)
#define AD7797_MODE_SEL(x)		(((x) & 0x7) << 13)	// Operation Mode Select
#define AD7797_MODE_CLKSRC(x)	(((x) & 0x3) << 6) 	// ADC Clock Source Select
#define AD7797_MODE_RATE(x)		((x) & 0xF) 		// Filter Update Rate Select

// AD7797_MODE_SEL(x) options
#define AD7797_MODE_CONT		 0 // Continuous Conversion Mode
#define AD7797_MODE_SINGLE	 1 // Single Conversion Mode
#define AD7797_MODE_IDLE		 2 // Idle Mode
#define AD7797_MODE_PWRDN		 3 // Power-Down Mode
#define AD7797_MODE_CAL_INT_ZERO 4 // Internal Zero-Scale Calibration
#define AD7797_MODE_CAL_INT_FULL 5 // Internal Full-Scale Calibration
#define AD7797_MODE_CAL_SYS_ZERO 6 // System Zero-Scale Calibration
#define AD7797_MODE_CAL_SYS_FULL 7 // System Full-Scale Calibration

// AD7797_MODE_CLKSRC(x) options
#define AD7797_CLK_INT		0 // Internal 64 kHz Clk not available at the CLK pin
#define AD7797_CLK_INT_CO	1 // Internal 64 kHz Clk available at the CLK pin
#define AD7797_CLK_EXT		2 // External 64 kHz Clock
#define AD7797_CLK_EXT_DIV2	3 // External Clock divided by 2

// AD7797_MODE_RATE(x) options
#define AD7797_RATE_123		  0x3 // 123Hz, no rejection
#define AD7797_RATE_62		  0x4 // 62Hz, no rejection
#define AD7797_RATE_50		  0x5 // 50Hz, no rejection
#define AD7797_RATE_33	    0x7 // 33.2Hz, no rejection
#define AD7797_RATE_16_80		0x9 // 16.7Hz, 80dB (60Hz only) rejection
#define AD7797_RATE_16_65		0xA // 16.7Hz, 65dB (50Hz and 60Hz) rejection
#define AD7797_RATE_12		  0xB // 12.5Hz, 66dB (50Hz and 60Hz) rejection
#define AD7797_RATE_10		  0xC // 10Hz, 69dB (50Hz and 60Hz) rejection
#define AD7797_RATE_8		    0xD // 8.33Hz, 70dB (50Hz and 60Hz) rejection
#define AD7797_RATE_6		    0xE // 6.25Hz, 72dB (50Hz and 60Hz) rejection
#define AD7797_RATE_4		    0xF // 4.17Hz, 74dB (50Hz and 60Hz) rejection

// Configuration Register Bit Designations (AD7797_REG_CONF)
#define AD7797_CONF_BO_EN	  (1 << 13) 			// Burnout Current Enable
#define AD7797_CONF_UNIPOLAR  (1 << 12) 			// Unipolar/Bipolar Enable
#define AD7797_CONF_CHAN(x)	  ((x) & 0x7) 			// Channel select


// AD7797_CONF_CHAN(x) options
#define AD7797_CH_AINP_AINM	0 // AIN(+) - AIN(-)
#define AD7797_CH_AINM_AINM	3 // AIN1(-) - AIN1(-)
#define AD7797_CH_TEMP			6 // Temp Sensor
#define AD7797_CH_AVDD_MONITOR	7 // AVDD Monitor

// ID Register Bit Designations (AD7797_REG_ID)
#define AD7797_ID			0x5B
#define AD7797_ID_MASK          0xFF

class PRDC_AD7797 {
  public:
    PRDC_AD7797();

    void setSPIFrequency(uint32_t);
    void setSPI(SPIClass&);

    bool begin();
    bool begin(uint8_t, uint8_t);
    void end();
    void reset();
    void setClockMode(uint8_t);
    void setRate(uint32_t);
    bool checkID();
    void waitReady();
    void setPower(uint8_t);
    void channelSelect(uint8_t);
    void calibrate(uint8_t, uint8_t);
    void setPolarity(uint8_t);
    uint32_t singleConversion();
    uint32_t continuousConversion();
    uint32_t continuousReadAverage(uint32_t);
    // float temperatureRead();
    // float rawToVolts(uint32_t, float);
    void printAllRegisters();

  private:
    SPISettings _spiSettings;
    SPIClass* _spi;
    const uint8_t _gain = 128;
    uint8_t _CS;
    uint8_t _MISO;
    uint8_t _polarity = 0;
    uint8_t _clock_mode = AD7797_CLK_INT;
    uint32_t _rate = AD7797_RATE_123;
    uint8_t _mode = AD7797_MODE_SINGLE;
    void pinInit();
    void beginTransaction();
    void endTransaction();
    uint32_t getRegister(uint8_t, uint8_t);
    uint32_t getSingleRegister(uint8_t, uint8_t);
    void setRegister(uint8_t, uint32_t, uint8_t);
    void setSingleRegister(uint8_t, uint32_t, uint8_t);
};
#endif	// _AD7797_H_
